src/server/engine_shard_set.cc

```diff
@@ -39,8 +49,260 @@ using absl::GetFlag;
+  }
+};
+
 vector<EngineShardSet::CachedStats> cached_stats;  // initialized in EngineShardSet::Init
 
+enum class MemDefragState : uint8_t {
+  REQUIRED,
+  NOT_REQUIRED,
+};
+
+enum class MemUtilizationState : uint8_t { UNKOWN, UNDER, OVER };
+
+// few issues here:
+// 1. Do we need to lock? (I think we do, since this is a task that I'm not sure is running
+// exclusively). - No we don't need a lock
+// 2. If we do need to lock, we can use the code similar to scan - ScanCb function. This require the
+// DB index, which we don't have! - we would be using index 0
+// 3. We need to find a way to map between the memory that we are moving to the new location that we
+// want to copy to
+//    we have a function that can tell use about memory usage for the given pointer (which in this
+//    case can be either the key or the value!), so are we going to move the value? (I think so),
+//    the key? (not sure), both?
+//  No we don't need, when we are doing new allocation, it would ensure that the allocation will
+//  happen to the first available space, this would result in moving all the memory to the "front"
+//  of the list.
+// 4, This leave the biggest issue, we need to find a new location to move to, how? - No we don't
+// see above To summarized:
+//  IF I understand this correctly:
+//    for each database id on this shared: (this is only 0)
+//      lock each key - don't
+//      check memory utilization for the address  of the key/value -> only for the value
+//      move the key/value to new location  -> just do allocation (for the value), move data to new
+//      memory location and free old one.
+//   The for each is running by the task itself, and the cursor is the for index - again for now
+//   only index 0!
+
+auto ReallocateValueImpl(PrimeValue& current_value) -> bool {
+  // we know that this is a string, we checked that before
+  auto val = current_value.ToString();
+  // we can now reset the old object as we are going to create a new one
+  current_value.Reset();
+  // This should allocate a new value
+  current_value.SetString(val);
+  return current_value.MallocUsed() > 0;
+}
+
+auto ReallocateValue(PrimeIterator it, void* from, const MemoryDefragParams& params, DbIndex index)
+    -> bool {
+  // allocate a new memory for the value, and then free the memory that used for the old one
+  CHECK(it->second.ObjType() == OBJ_STRING);  // only supporting for string right now!
+  if (it->second.IsExternal()) {
+    VLOG(1) << std::hex << from << ": the object that we are dealing with is external for type "
+            << std::dec << OpTypeToString(params.object_type)
+            << " or was not allcoated on the heap";
+    return false;
+  }
+
+  params.slice->PreUpdate(index, it);
+  auto key = it->first.ToString();
+  auto state = ReallocateValueImpl(it->second);
+  params.slice->PostUpdate(index, it, key);
+  return state;
+}
+
+auto MemoryReallocate(PrimeIterator it, const MemoryDefragParams& params, DbIndex index)
```

Comment:
why using `auto` is better then just specify a return type? I prefer avoiding this notation, frankly.

Reply:
Personally I think this is more readable - maybe this is because all other languages today using it, I saw that you're using it as well. If this is an issue I can change this to the more traditional C style.  again for me its very hard with the old way of (return type) (function name)(type val..), but this is not a religion for me

Reply:
I use it only when I need to return an inner type. 
"auto" word does not provide any meaningful information, so it's just a fluff in this case.